### ISSUE TITLE
Reloading WMTS layer in OpenLayers 6.4.3 and newer

### ISSUE THREAD

**Issue by petrgon** [openlayers/openlayers#13994]
Hello, I am sorry for double posting this. I already asked on [StackOverflow ](https://stackoverflow.com/questions/73237926/reloading-wmts-layer-in-openlayers-6) but I did not get any answer. So I am trying my luck here because it seems to me as this could be actually a bug. I am just kindly asking you to check my trying if I am doing something wrong or confirming if it is a bug, please.

**Describe the bug**
I'm trying to reload the WMTS layer after changing some tiles on the remote WMTS server.

The problem is that I am not able to reload tiles in the current viewport (or in the current zoom level). Other lower resolution tiles are getting reloaded correctly. The issue is that tiles in the viewport redraw but a request to verify ETag is not made. A suspicious thing is that if I disable cache in browser's network tab tiles in the viewport are getting requested correctly.

I am using following snippet for cache invalidation, I do have more layers of different type:

```js
const source = layer.getSource(); // WMTS layer, returns WMTS source

if (source != null) {
  source.getTileCacheForProjection?.(this.map.getView().getProjection())?.clear();
  source.clear();
  source.tileCache.clear();
  source.refresh?.();
  source.changed();
}

layer.changed();
```

I tried the snippet in OL v6.3.1 where it worked correctly but not in 6.4.3, 6.14.1 (I use in master branch), or 6.15.1.

My WMTS server does use ETag. Here is two examples of the response header with ETag set

```
HTTP/1.1 200 OK
Server: Microsoft-HTTPAPI/2.0
Access-Control-Allow-Origin: *
Cache-Control: private, max-age=0, must-revalidate
Content-Disposition: inline; filename=ows-data.png
Content-Length: 64795
Content-Type: image/png
ETag: a22c283e43ab4143
X-Frame-Options: SAMEORIGIN
Date: Thu, 04 Aug 2022 14:20:00 GMT
```
```
HTTP/1.1 304 Not Modified
Server: Microsoft-HTTPAPI/2.0
Date: Thu, 04 Aug 2022 14:33:56 GMT
```
Server correctly returns 304 for lower resolution tiles and 200 with payload for those that changed.

**Expected behavior**
Cache will invalidate properly and ETag is properly revalidated for tiles in viewport.

**Comment by petrgon** [comment]
Ahocevar, thank you for the quick response. Yes, that is my case, I have the WMTS layer reprojected. I've tried your snippet but no success.

Can I help you fix this bug? With my manager approval, I can contribute to the source code, it would be nice if you could direct me where to start looking. Or I can help with testing at least, I have the environment ready.

**Comment by ahocevar** [comment]
@petrgon Can you please try the fix in #14013? According to the tests I added there, it should fix your problem.

**Comment by petrgon** [comment]
Thank you Ahocevar for the try, but I am sorry, it does not work. I was testing it primarily using your snippet `layer.getSource().refresh()`.

It is improvement, originally when I did `layer.getSource().refresh();` nothing re-rendered. Now it does something. I can see tiles in viewport being unloaded (I can see gray background color) and loaded again. But no ETag validation request for tiles in viewport is made.

I can see some requests being made for low-res tiles. But the behavior in nondeterministic. When I call `layer.getSource().refresh();` multiple times I can see the same repeating number of requests for ETag validation being made (e.g. 45-3-0, it depends on zoom level). 

As I said, these tiles are low resolution tiles and they contains correct map data (I am switching between satellite tiles with snow and without snow). But no request is being made for tiles in viewport they always keep the same amount of snow level as before.

**Comment by ahocevar** [comment]
@petrgon Are you sure you tested the correct version, i.e. v7.0.0?

Also, just to rule out the chance that your browser caches more than you think (which I think it does, because your server does not send a `Last-Modified:` header): what happens when you clear the browser cache before calling refresh()? In that case, you should see all tiles getting refetched. And if that's the case, you'd have to double-check the cache related headers from your server.

You can try it out here: https://codesandbox.io/s/scaleline-indiana-east-forked-7hdii8?file=/index.html. Also note that when you change the ol dependency's version from v7.0.0 to v6.15.1, no request will be made upon refresh() at all, which was the probem I fixed.

**Comment by petrgon** [comment]
Thank you @ahocevar for your help. I cloned the OL repo and checkouted the master, I verified that the commit [fae9e90](https://github.com/openlayers/openlayers/commit/fae9e9097d16969c2a732d18a142a9418a0d3653) is there. Then I built it using  `npm run build-package` and installed it using `npm install w:\external\openlayers\build\ol\` (path to the build output). As a confirmation that I used correct branch was the need of updating `ol-mapbox-style`. In the `packaged.json`, I can see that the `ol` package targets to the local filesystem. The project works fine (except this issue) after all these steps.

✔️ I can also confirm that something did change. Calling `layer.getSource().refresh();` gives me a visual feedback that it tries to refetch the map data. It didn't before your commit.

⚠️ The issue is that no HTTP request is made to validate the ETag. Tiles that are loaded after `layer.getSource().refresh();` are the same as before. I don't know where these tiles come from. Unfortunately, I don't even know how to find out. I agree that this doesn't have to be an issue of OL anymore.

I'll do more investigation on my side. I'll be in touch. Thank you again for your cooperation.
